Change: Merge Game Options and Game Settings together.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

We sometimes talk about revamping the way game options and settings are done, but it's always been ideas for reworking it completely instead of just doing it.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Merge Game Options and Settings into one window. Game Settings is now an additional 'Advanced' tab in the Game Options window.

It looks a bit weird being switched from mauve to grey. An alternative is switching the grey to mauve, to match the (still separate) AI, GS and NewGRF Settings windows.

![image](https://github.com/user-attachments/assets/47c189d0-7ead-4284-8a20-8530acc815b7)

The setting description box is quite large, due to some settings having an essay describing them.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

PeterN committed:
Change: Merge Game Options and Game Settings together.

Game Settings is now an additional 'Advanced' tab in the Game Options window.

rubidium42 commented on src/settings_gui.cpp:
```diff
@@ -661,6 +820,72 @@ struct GameOptionsWindow : Window {
+
+	void OnPaint() override
+	{
+		if (this->GetWidget<NWidgetStacked>(WID_GO_TAB_SELECTION)->shown_plane != 3) {
```
What is so special about the socials tab that it needs a new `OnPaint` implementation? Or should this have been a 4? If so, might having an enumeration be a better solution going forward?

Also, when searching (i.e. have some text in the advanced tab search filter) in a 'Category' other than 'Expert', the notice is shown that 'some search results are hidden' on the social tab.
![image](https://github.com/user-attachments/assets/d2a5e987-254b-481c-be35-da64b523f0c1)

PeterN replied:
Whoops, yes, definitely should be 4. I experimented with order and didn't fix it.

rubidium42 commented on src/settings_internal_gui.cpp:
```diff
@@ -0,0 +1,957 @@
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file settings_internal_gui.cpp GUI for settings. */
```
This comment is the same as for settings_gui.cpp. Maybe mention what's specifically supposed to be in here?

rubidium42 commented on src/settings_internal_gui.h:
```diff
@@ -0,0 +1,172 @@
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file settings_gui.h Functions for setting GUIs. */
```
The filename's wrong. The remainder of this comment is the same as for settings_gui.h. Maybe mention what's specifically supposed to be in here?

PeterN committed:
Change: Use standard mauve/yellow scheme for game options window.

This is the colour scheme used by AI/GS settings, NewGRF settings and (previously) advanced settings.

PeterN committed:
Change: Slider widget can now be colourful.

frosch123 commented on src/settings_gui.cpp:
```diff
@@ -357,26 +357,95 @@ std::unique_ptr<NWidgetBase> MakeNWidgetSocialPlugins()
 	return std::make_unique<NWidgetSocialPlugins>();
 }
 
+int SETTING_HEIGHT = 11;    ///< Height of a single setting in the tree view in pixels
```
Someone knows why this is no constant in some header? It's never assigned.

PeterN replied:
It is assigned, hidden in a bad way...

`resize.height = SETTING_HEIGHT = std::max({(int)_setting_circle_size.height, SETTING_BUTTON_HEIGHT, GetCharacterHeight(FS_NORMAL)}) + WidgetDimensions::scaled.vsep_normal;`

The `= 11;` is complete distraction and wrong.

PeterN committed:
Change: Add alternate_colour widget data for boolean button.

Allows boolean button to have different background and button colours.

PeterN committed:
Codechange: Remove unnecessary SetMinimalSize from Game Options window.

PeterN committed:
Change: Adjust section names in Graphics options tab.

"Interface size" and "Graphics" section names do not make much sense any more.

2TallTyler reviewed (dismissed):
One last question to wonder aloud, do we take this opportunity to standardize on our use of the word Setting or Option? Since we are about to make players re-learn where to look for things? We have previously needed both words but now could pick one to standardize on.

The current status is:

#### Options:
* Game options
* Sandbox options
* Transparency options
#### Settings:
* NewGRF settings (which have a bonus word, "parameters")
* AI settings (which also have parameters, although the button says `Configure`)
